File: DevelopArchitecture/DevelopArchitecture/threeclass/photo/JMainPhotoPickerCell.m
//
//  JMainPhotoPickerCell.m
//  DevelopArchitecture
//
//  Created by Jeffrey on 2019/4/1.
//  Copyright © 2019 Jeffrey. All rights reserved.
//

#import "JMainPhotoPickerCell.h"
#import "JPhotoPickerInterface.h"
@implementation JMainPhotoPickerCell{
    UIImageView *headImageView;
    UILabel *titleLabel;
    JPhotosDataModel *m_dataModel;
}
-(void)funj_addBaseTableSubView{
    self.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    headImageView =[UIImageView funj_getImageView:CGRectMake(10, 10, (90+IS_IPAD*40), kImageViewHeight(90+IS_IPAD*40)) img:@""];
    headImageView.layer.masksToBounds = YES;
    headImageView.contentMode = UIViewContentModeScaleAspectFill;
    [self.contentView addSubview:headImageView];
    titleLabel =[UILabel funj_getLabel:CGRectMake(headImageView.right+10, headImageView.top, kphotoPickerViewWidth-headImageView.right-40, headImageView.height) fc:JTextFCMake(kFont_Size15, kColor_Text_Black_Dark)];
    [self.contentView addSubview:titleLabel];
    
    UIImageView *line =[UIImageView funj_getLineImageView:CGRectMake(headImageView.right+10, headImageView.bottom+9, kphotoPickerViewWidth-30-headImageView.right-10, 1)];
    [self.contentView addSubview:line];
}
-(void)funj_setBaseTableCellWithData:(JPhotosDataModel *)data{
    m_dataModel = data;
    NSMutableAttributedString *nameString = [[NSMutableAttributedString alloc] initWithString:data.name attributes:@{NSFontAttributeName:[UIFont systemFontOfSize:16],NSForegroundColorAttributeName:kColor_Text_Black_Dark}];
    NSAttributedString *countString = [[NSAttributedString alloc] initWithString:[NSString stringWithFormat:@"  (%zd)", data.count] attributes:@{NSFontAttributeName:[UIFont systemFontOfSize:16], NSForegroundColorAttributeName:[UIColor lightGrayColor]}];
    [nameString appendAttributedString:countString];
    titleLabel.attributedText = nameString;
    LRWeakSelf(self);
    [JPhotoPickerInterface funj_getPhotoWithAsset:[m_dataModel.fetchResult lastObject] type:PHImageRequestOptionsDeliveryModeFastFormat photoWidth:80 completion:^(UIImage * _Nonnull image, NSDictionary * _Nonnull dic, BOOL isDegraded) {
        LRStrongSelf(self);
        self->headImageView.image = image;
    }];
    
    
}

@end

@implementation JPhotoPickerCell{
    UIImageView *coverImageView;
    UIButton *selectBt;
    UILabel *countLabel;
    UIButton*timeLabel;
    JPhotoPickerModel *m_dataModel;
}
-(void)funj_addBaseCollectionView{
    CGFloat width =  self.width;

    coverImageView =[UIImageView funj_getImageView:CGRectMake(0,0,width, kImageViewHeight(width)) img:@""];
    coverImageView.contentMode = UIViewContentModeScaleAspectFill;
    [self.contentView addSubview:coverImageView];
    coverImageView.layer.masksToBounds = YES;
    
    //VideoSendIcon
    if([JPhotosConfig share].m_currentIsVideo){
        timeLabel =[UIButton funj_getButtons:CGRectMake(30, coverImageView.height-20, coverImageView.width-40, 15) t:@"00:00"  fc:JTextFCMake(kFont_Size10, [UIColor grayColor]) img:@[@"VideoSendIcon"] d:nil  a:nil  tag:0 set:^(UIButton *button) {
            [button funj_updateContentImageLayout:kRIGHT_IMAGECONTENT a:JAlignMake(0, 10, 0)];
        }];
        [self.contentView addSubview:timeLabel];
    }
    
    selectBt =[UIButton funj_getButtons:CGRectMake(coverImageView.width-60, 0, 60, 60) t:nil  fc:JTextFCZero() img:@[@"photo_def_photoPickerVc",@"photo_sel_photoPickerVc"] d:self  a:@"funj_selectToAdd:" tag:0 set:nil];
    [self.contentView addSubview:selectBt];
    selectBt.contentHorizontalAlignment = UIControlContentHorizontalAlignmentRight;
    selectBt.contentVerticalAlignment = UIControlContentVerticalAlignmentTop;
    [selectBt funj_updateButtonSelectStyle:NO  ischange:NO];
    
    if([JPhotosConfig share].m_isMultiplePhotos){
        countLabel =[UILabel funj_getOneLabel:CGRectMake(5, coverImageView.height-20, 15, 15) fc:JTextFCMakeAlign(kFont_Size10, kColor_White,NSTextAlignmentCenter)];
        [self.contentView addSubview:countLabel];
        [countLabel funj_setViewCornerLayer:JFilletMake(0, 15/2, kColor_Clear)];
        countLabel.backgroundColor = [UIColor redColor];
    }
}
-(void)funj_setBaseCollectionData:(JPhotoPickerModel *)data{
    m_dataModel = data;
    [timeLabel setTitle:data.timeLength forState:UIControlStateNormal];
    [timeLabel funj_updateContentImageLayout:kRIGHT_IMAGECONTENT a:JAlignMake(0, 10, 0)];
    countLabel.text = [NSString stringWithFormat:@"%zd",data.indexCount];
    selectBt.selected = data.isSelected;
    countLabel.hidden = data.indexCount <=0;
    LRWeakSelf(self);
    [JPhotoPickerInterface funj_getPhotoWithAsset:data.asset type:PHImageRequestOptionsDeliveryModeFastFormat photoWidth:coverImageView.width completion:^(UIImage * _Nonnull image, NSDictionary * _Nonnull dic, BOOL isDegraded) {
        LRStrongSelf(self);
        self->coverImageView.image = image;
    }];
    
}

-(void)funj_reloadCountIndex:(NSInteger)index{
    countLabel.text =[NSString stringWithFormat:@"%zd",index];
    countLabel.hidden = index <=0;
}
-(void)funj_selectToAdd:(UIButton*)sender{
    sender.selected = !sender.selected;
    m_dataModel.isSelected = sender.selected;
    if(self.m_selectItemCallback)self.m_selectItemCallback(sender,m_dataModel);
}
@end

File: DeveloperArchitecture/DeveloperArchitecture/threeclass/photo/cell/JMainPhotoPickerCell.swift
//
//  JMainPhotoPickerCell.swift
//  DeveloperArchitecture
//
//  Created by Jeffrey on 2020/11/7.
//  Copyright © 2020 Jeffery. All rights reserved.
//

import Foundation
import UIKit

class JMainPhotoPickerCell : JBaseTableViewCell {
    var m_imageView : UIImageView?
    var m_titleLabel : UILabel?
    var m_lineImageView : UIImageView?
    
    override func funj_addBaseTableSubView() {
        self.accessoryType = .disclosureIndicator
        
        m_imageView = UIImageView(i: CGRect(x: 20, y: 10, width: 90, height: kImageViewHeight(100 + kis_IPad_1 * 20 )), image: nil)
        self.contentView.addSubview(m_imageView!)
        m_imageView?.contentMode = .scaleAspectFill
        m_imageView?.layer.masksToBounds = true
        
        m_titleLabel = UILabel(i: CGRect(x:m_imageView!.right + 20 , y: m_imageView!.center.y - 10, width: kWidth - m_imageView!.width - 40, height: 20), title: "", textFC: JTextFC(f: kFont_Size17, c: kColor_Text_Black))
        self.contentView.addSubview(m_titleLabel!)
        
        m_lineImageView = UIImageView(i_line: CGRect(x: m_titleLabel!.left, y: m_imageView!.bottom + 9 , width: kWidth - m_imageView!.right - 20 * 2, height: 1))
        self.contentView.addSubview(m_lineImageView!)
    }
    override func funj_setBaseTableCellWithData(_ data: Any) {
        let model = data as? JPhotosDataModel
        let nameString = NSMutableAttributedString(string: model!.m_name ?? "", attributes: [NSAttributedString.Key.font : kFont_Size16 , NSAttributedString.Key.foregroundColor : kColor_Text_Black_Dark])
        let countStr = NSMutableAttributedString(string: String(format: "  (%d)", model!.m_count), attributes: [NSAttributedString.Key.font : kFont_Size16 , NSAttributedString.Key.foregroundColor : kColor_Text_GRAY_Dark])
        nameString.append(countStr)

        m_titleLabel?.attributedText = nameString
        
        JPhotoPickerInterface.funj_getPhotoWithAsset(phAsset: model?.m_fetchResult?.lastObject, deliveryMode: .fastFormat, width: 80) { [weak self](image, dic, isDegraded) in
            self?.m_imageView?.image = image
        }
        
    }
}
